pkg/executor/sortexec/sort.go

```diff
@@ -119,57 +122,173 @@ func (e *SortExec) Next(ctx context.Context, req *chunk.Chunk) error {
+
+	if e.helper.spillError != nil {
+		return e.helper.spillError
+	}
+
+	if e.sortPartitions[0].isSpillTriggered() {
+		if len(e.cursors) == 0 {
+			e.initCursors()
+		}
+
+		// Maybe the spill is triggered by the last chunk, so there is only one partition.
+		cursor := e.cursors[0]
+		for !req.IsFull() {
+			row := cursor.getSpilledRow()
+			if row == nil {
+				success, err := e.reloadCursor(0)
+				if err != nil {
+					return err
+				}
+
+				if !success {
+					// All data have been completely consumed
+					return nil
+				}
+
+				// row shouldn't be nil here
+				row = cursor.getSpilledRow()
 			}
-			e.Idx++
+
+			cursor.advanceRow()
+			req.AppendRow(*row)
+		}
+	} else {
+		// Spill is not triggered and all data are in memory.
+		rowNum := e.sortPartitions[0].numRowInMemory()
+		for !req.IsFull() && e.sortPartitions[0].getIdx() < rowNum {
+			e.sortPartitions[0].getSortedRowFromMemoryAndAppendToChunk(req)
 		}
 	}
 	return nil
 }
 
 func (e *SortExec) externalSorting(req *chunk.Chunk) (err error) {
 	if e.multiWayMerge == nil {
-		e.multiWayMerge = &multiWayMerge{e.lessRow, e.compressRow, make([]partitionPointer, 0, len(e.PartitionList))}
-		for i := 0; i < len(e.PartitionList); i++ {
-			chk := chunk.New(exec.RetTypes(e), 1, 1)
+		err := e.initExternalSorting()
+		if err != nil {
+			return err
+		}
+	}
 
-			row, _, err := e.PartitionList[i].GetSortedRowAndAlwaysAppendToChunk(0, chk)
+	for !req.IsFull() && e.multiWayMerge.Len() > 0 {
+		// Get and insert data
+		element := e.multiWayMerge.elements[0]
+		req.AppendRow(element.row)
+
+		// Get a new row from that partition which inserted data belongs to
+		partitionID := element.partitionID
+		newRow := e.cursors[partitionID].getSpilledRow()
+		if newRow == nil {
+			success, err := e.reloadCursor(partitionID)
 			if err != nil {
 				return err
 			}
-			e.multiWayMerge.elements = append(e.multiWayMerge.elements, partitionPointer{chk: chk, row: row, partitionID: i, consumed: 0})
+			if !success {
+				// All data in this partition have been consumed
+				heap.Remove(e.multiWayMerge, 0)
+				continue
+			}
+			newRow = e.cursors[partitionID].getSpilledRow()
 		}
-		heap.Init(e.multiWayMerge)
+
+		e.cursors[partitionID].advanceRow()
+
+		e.multiWayMerge.elements[0].row = *newRow
+		heap.Fix(e.multiWayMerge, 0)
 	}
+	return nil
+}
 
-	for !req.IsFull() && e.multiWayMerge.Len() > 0 {
-		partitionPtr := e.multiWayMerge.elements[0]
-		req.AppendRow(partitionPtr.row)
-		partitionPtr.consumed++
-		partitionPtr.chk.Reset()
-		if partitionPtr.consumed >= e.PartitionList[partitionPtr.partitionID].NumRow() {
-			heap.Remove(e.multiWayMerge, 0)
-			continue
+func (e *SortExec) switchToNewSortPartition(fields []*types.FieldType, byItemsDesc []bool) {
+	// Put the full partition into list
+	e.sortPartitions = append(e.sortPartitions, e.partition)
+
+	e.partition = newSortPartition(fields, e.MaxChunkSize(), byItemsDesc, e.keyColumns, e.keyCmpFuncs, e.helper, e.spillLimit)
```

Comment:
```suggestion
	e.curPartition = newSortPartition(fields, e.MaxChunkSize(), byItemsDesc, e.keyColumns, e.keyCmpFuncs, e.helper, e.spillLimit)
```
rename to `curPartition`?

Reply:
okk